### PR TITLE
[Improvement] Split basic checks in CI pipeline

### PR DESCRIPTION
**Pull request by kaijchen**
### What changes were proposed in this pull request?

1. Enable CI on non-master branches.
2. Split `checkstyle` and `rat` checks from `build` matrix.
3. Add `spotbugs`, its result is ignored since there are many violations right now.
4. Add `Summary of failures` step in each job.

### Why are the changes needed?

1. Contributors can ensure CI has passed on forked branches before submitting PR.
2. It's easier to see why CI has failed. And avoids repeating `checkstyle` and `rat`.
3. Spotbugs serves as suggestions for now.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Example of `rat` failure: https://github.com/kaijchen/Firestorm/actions/runs/2383526195

### CODE REVIEW

**Comment by kaijchen** [comment]
Seems CI for draft PR is running anyway, let's just use the default configs.

**Comment by kaijchen** [comment]
Marking this PR as draft to work on more CI improvements.

**Comment by kaijchen** [comment]
@jerqi PTAL.

File: checks/rat.sh [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+# License at
+#
+# https://opensource.org/licenses/Apache-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations under the License.
+
+mvn -B -fae org.apache.rat:apache-rat-plugin:check
```
[the remark this answers is not in the document]
Comment (kaijchen) [reply checks/rat.sh:19]: Thanks for the correction.

[the remark this answers is not in the document]
Comment (kaijchen) [reply checks/rat.sh]: `checkstyle` and `rat` don't compile the code. So actually this works.

However, maybe it's better to put them in separate matrix then.

[the remark this answers is not in the document]
Comment (jerqi) [reply checks/rat.sh]: checks/${{ matrix.check }}.sh -Pspark2 && checks/${{ matrix.check }}.sh -Pspark3 && checks/${{ matrix.check }}.sh -Pmr？

[the remark this answers is not in the document]
Comment (kaijchen) [reply checks/rat.sh]: I've found a better way to do it. I'll update this PR later.

[the remark this answers is not in the document]
Comment (jerqi) [reply checks/rat.sh]: OK

[the remark this answers is not in the document]
Comment (jerqi) [reply checks/rat.sh]: Can we have one pipeline to check all profiles style?

[the remark this answers is not in the document]
Comment (kaijchen) [reply checks/rat.sh]: > Can we have one pipeline to check all profiles style?

It's easier to maintain this way. Say if in the future we added a new profile `flink`.
We only need to add one line in `maven.yml`.

[the remark this answers is not in the document]
Comment (kaijchen) [reply checks/rat.sh]: However, there is a way to do it while keeping maintainability as demonstrated in 44ac493233d504f88feafd25317decad468e8f5b.

[the remark this answers is not in the document]
Comment (jerqi) [reply checks/rat.sh]: I thinkt it's a better way.

**Comment by frankliee** [comment]
image
There seems too many CI pipelines...
There are also -Pspark2.3 -Pspark3.1 -Pspark3.1  -Pspark3.0

**Comment by kaijchen** [comment]
> There seems too many CI pipelines... There are also -Pspark2.3 -Pspark3.1 -Pspark3.1 -Pspark3.0

Seems these profiles were not tested in the existing CI.

**Comment by frankliee** [comment]
> > There seems too many CI pipelines... There are also -Pspark2.3 -Pspark3.1 -Pspark3.1 -Pspark3.0
> 
> Seems these profiles were not tested in the existing CI.

Yes, I think there will be another PR to fix this since we have parallelized pipelines.

**Comment by kaijchen** [comment]
> There seems too many CI pipelines... There are also -Pspark2.3 -Pspark3.1 -Pspark3.1 -Pspark3.0

Adding them is as simple as a0034f90fc4c6eb0037c732783cddd54bd34006c. It's not too messy since they are grouped.

05a0a426-b47c-46a5-9319-c7eac3eaa749

**Comment by jerqi** [comment]
@kaijchen Please not add the profile 3.0 and profile 2.3.in another pr, there seems to be something wrong with profile 3.0.  @frankliee @duanmeng Could you fix the compile issue with profile 3.0.

**Comment by jerqi** [comment]
https://github.com/Tencent/Firestorm/pull/154
We fix profile spark 3.0

**Comment by kaijchen** [comment]
> We fix profile spark 3.0

Good fix. You can add new profiles in `.github/workflows/parallel.yml` and `.github/workflows/sequential.yml`.

**Pull request merged**